62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+	}
+
+	return result;
+}
+
+Hatch::Hatch(core::SRange<CPolyline> lines, const MajorAxis majorAxis, int32_t& debugStep, std::function<void(CPolyline, CPULineStyle)> debugOutput /* tmp */)
+{
+    std::stack<Segment> starts; // Next segments sorted by start points
+    std::stack<double> ends; // Next end points
+    double maxMajor;
+
+    int major = (int)majorAxis;
+    int minor = 1-major; // Minor = Opposite of major (X)
+
+    auto drawDebugBezier = [&](QuadraticBezier bezier, float32_t4 color)
+    {
+        CPolyline outputPolyline;
+        std::vector<QuadraticBezierInfo> beziers;
+        QuadraticBezierInfo bezierInfo;
+        bezierInfo.p[0] = bezier.p[0];
+        bezierInfo.p[1] = bezier.p[1];
+        bezierInfo.p[2] = bezier.p[2];
+        beziers.push_back(bezierInfo);
+        outputPolyline.addQuadBeziers(core::SRange<QuadraticBezierInfo>(beziers.data(), beziers.data() + beziers.size()));
+
+        CPULineStyle cpuLineStyle;
+        cpuLineStyle.screenSpaceLineWidth = 4.0f;
+        cpuLineStyle.worldSpaceLineWidth = 0.0f;
+        cpuLineStyle.color = color;
+
+        debugOutput(outputPolyline, cpuLineStyle);
+    };
+
+    auto drawDebugLine = [&](float64_t2 start, float64_t2 end, float32_t4 color)
+    {
+        CPolyline outputPolyline;
+        std::vector<float64_t2> points;
+        points.push_back(start);
+        points.push_back(end);
+        outputPolyline.addLinePoints(core::SRange<float64_t2>(points.data(), points.data() + points.size()));
+        
+        CPULineStyle cpuLineStyle;
+        cpuLineStyle.screenSpaceLineWidth = 2.0f;
+        cpuLineStyle.worldSpaceLineWidth = 0.0f;
+        cpuLineStyle.color = color;
+        
+        debugOutput(outputPolyline, cpuLineStyle);
+    };
+
+    {
+        std::vector<Segment> segments;
+        for (CPolyline& line : lines)
+        {
+            for (uint32_t secIdx = 0; secIdx < line.getSectionsCount(); secIdx ++)
+            {
+				auto addBezier = [&](QuadraticBezier bezier)
+				{
+					auto outputBezier = bezier;
+					if (outputBezier.evaluateBezier(0.0)[major] > outputBezier.evaluateBezier(1.0)[major])
+					{
+						outputBezier.p[2] = bezier.p[0];
+						outputBezier.p[0] = bezier.p[2];
+						assert(outputBezier.evaluateBezier(0.0)[major] <= outputBezier.evaluateBezier(1.0)[major]);
+					}
+
+					//if (debugOutput)
+					//	drawDebugBezier(outputBezier, float32_t4(0.0, 0.0, 0.0, 1.0));
+					beziers.push_back(outputBezier);
+				};
+
+                auto section = line.getSectionInfoAt(secIdx);
+                if (section.type == ObjectType::LINE)
+                    {
+						for (uint32_t itemIdx = section.index; itemIdx < section.index + section.count; itemIdx++)
+						{
+							auto begin = line.getLinePointAt(itemIdx);
+							auto end = line.getLinePointAt(itemIdx + 1);
+							addBezier({ 
+								// TODO: it was suggested at some point the middle point should be NaN for this,
+								// should we go ahead with that instead?
```

Comment:
The point was to detect a line better so IsLineSegment will be `whatYouCheckedBefore || isnan(p1)` 
but I think you're fine now, it doesn't matter. remove todo